### PR TITLE
Meshmoving super imposing on existing motion

### PR DESCRIPTION
**Pull request by adityaghantasala**
Currently the mesh moving application overwrites any previously defined movement and the mesh velocity calculated and set previously. 

For example, I there is a rotation motion done via a process, then one want to FSI, one has to also use the mesh motion app to calculate the mesh deformation. In this case, the current behaviour is that the current mesh is overwritten (see changes) by the mesh motion solve. 

The current changes propose a way to overcome this by superimposing the displacement and velocity calculated by MeshMotionApp on the existing ones. 

This is straight forward to do with MESH_VELOCITY. But since MESH_DISPLACEMENT is the DOF for MeshMotionApp it is not straight forward, as the app sets the current to initial and then updates them with MESH_DISPLACEMENT. 

In the changes I made, it uses the DISPLACEMENT variable to transfer the displacement from an external motion. This approach may not be the best. And I am open to changes here.

### CODE REVIEW

**Comment by philbucher** [comment]
hm this breaks all tests ...

could the same that you do with mesh-disp also be done with mesh-vel?

**Comment by adityaghantasala** [comment]
@philbucher 

> could the same that you do with mesh-disp also be done with mesh-vel?


did not get you ! 

Also ya I can expect it breaks the tests as they might not consider the +=

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
The FM-ALE utilities use the `DISPLACEMENT` as boundary condition for the L2 minimization problem that computes the `MESH_DISPLACEMENT` fixed nodes values. Doing this would apply a wrong displacement values when moving the mesh after solving the ALE problem.

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ void MoveMesh(const ModelPart::NodesContainerType& rNodes) {
     for (int i=0; i<num_nodes; i++) {
         const auto it_node  = nodes_begin + i;
         noalias(it_node->Coordinates()) = it_node->GetInitialPosition()
-            + it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT);
+            + it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT) + it_node->FastGetSolutionStepValue(DISPLACEMENT);
```
Comment (philbucher) [inline applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp]:
```suggestion
            + it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT) + it_node->GetValue(AUX_MESH_DISPLACEMENT);
```

File: applications/MeshMovingApplication/custom_utilities/mesh_velocity_calculation.cpp [hunk 2]
```diff
@@ -86,7 +86,7 @@ void CalculateMeshVelocities(ModelPart& rModelPart,
     for (int i=0; i<num_local_nodes; i++) {
         const auto it_node  = nodes_begin + i;
         auto& r_mesh_v0 = it_node->FastGetSolutionStepValue(MESH_VELOCITY);
-        noalias(r_mesh_v0)  = coeffs[0] * it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT);
+        noalias(r_mesh_v0) += coeffs[0] * it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT);
         noalias(r_mesh_v0) += coeffs[1] * it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT, 1);
         noalias(r_mesh_v0) += coeffs[2] * it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT, 2);
     }
```
Comment (philbucher) [inline applications/MeshMovingApplication/custom_utilities/mesh_velocity_calculation.cpp:93]:
```suggestion
noalias(r_mesh_v0) += it_node->GetValue(AUX_MESH_VELOCITY);
        }
```

note that here we would need to check before (or access the Node as `const`) if the `AUX_MESH_VELOCITY` exists

**Review by philbucher — COMMENTED** [review]
I hope these suggestions help to understand what I meant.
Would what I propose work for you @adityaghantasala ?
- I would not make the additinoal vars historical to save memory (and because allocating hist-vars is a bit of a pain, so I would only do it if necessary)
- The naming of the vars is up for discussion ofc

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp (hunk 1 above)
Comment (adityaghantasala) [reply applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp]: Can we specify which variable to use here (from JSON) ?

Comment (philbucher) [reply applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp]: Would this work for you?

In principle yes, but this would propagate quite a bit throughout the code.
Hence I would prefer to set it to a specific variable (I mean, also other variables are specific, we cannot just use what we want)

**Comment by rubenzorrilla** [comment]
I also agree with having a new variable for these purposes (we can discuss about the name). I also share @philbucher 's opinion regarding the memory overhead of having an historical variable for this purpose. Aside of the fact that using a non-historical one makes it possible to use the variable without requiring to allocate it outside of the mesh moving util.

**Comment by adityaghantasala** [comment]
Having the variable as non-historical also fine for me and makes sense.
Only thing is about what do we call the variable and if it should be fixed (the name). Is this general enough to fix the name.

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp (hunk 1 above)
Comment (adityaghantasala) [reply applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp]: Yes that works I believe. 

I did not understand by "other variables are specific" ? Do you mean the MESH_VELOCITY ?

**Comment by rubenzorrilla** [comment]
> Having the variable as non-historical also fine for me and makes sense.
> Only thing is about what do we call the variable and if it should be fixed (the name). Is this general enough to fix the name.

Although having a fixed name would imply that the user needs to copy the auxiliary displacement from the variable in which it is stored to the `NEW_MESH_DISPLACEMENT` var, I'd rather go for a fixed name (don't know which).

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp (hunk 1 above)
Comment (philbucher) [reply applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp]: With other variables I mean things like VELOCITY/PRESSURE for fluid or DISPLACEMENT/ROTATION in structural

Those are fixed

**Comment by philbucher** [comment]
you could try passing it through json, but as said before I think you it will propagate a lot through the code

**Comment by adityaghantasala** [comment]
Ok .. for now lets got with AUX_* as Phillip suggests, but then these should go into core so as to be able to be accessible from all places right ?

**Comment by philbucher** [comment]
Yes I think so, unless one would only use it from python

**Comment by adityaghantasala** [comment]
@philbucher  the current version of of the branch has the changes as we discussed. 
@rubenzorrilla  now the changes do not change the previous behaviour. The superimposition is configurable via python and will not effect the existing behaviour.

File: applications/MeshMovingApplication/python_scripts/ale_fluid_solver.py [hunk 3]
```diff
@@ -190,11 +191,18 @@ def SolveSolutionStep(self):
         for mesh_solver in self.mesh_motion_solvers:
             is_converged &= mesh_solver.SolveSolutionStep()
 
+        for mesh_solver in self.mesh_motion_solvers:
+            KMM.MoveMesh(
```
Comment (philbucher) [inline applications/MeshMovingApplication/python_scripts/ale_fluid_solver.py]: this is a leftover, please remove it

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.h [hunk 4]
```diff
@@ -36,6 +36,13 @@ void MoveMesh(const ModelPart::NodesContainerType &rNodes);
 ModelPart* GenerateMeshPart(ModelPart &rModelPart,
                                     const std::string &rElementName);
 
+void SuperImposeVariables(ModelPart &rModelPart, const Variable< array_1d<double, 3> >& rVariable,
+                                                 const Variable< array_1d<double, 3> >& rVariableToSuperImpose);
+
+void SuperImposeMeshDisplacement(ModelPart &rModelPart, const Variable< array_1d<double, 3> >& rVariableToSuperImpose);
+
+void SuperImposeMeshVelocity(ModelPart &rModelPart, const Variable< array_1d<double, 3> >& rVariableToSuperImpose);
```
Comment (philbucher) [inline applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.h:44]: are those necessary?
You could pass directly the first variable too

**Review by philbucher — CHANGES_REQUESTED** [review]
don't forget the trilinos strategies!

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.h (hunk 4 above)
Comment (adityaghantasala) [reply applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.h:44]: strictly speaking not required but I liked this way because its more readable when using those functions.

**Comment by adityaghantasala** [comment]
The trillinos ones are just using the base solver SolveSolutionStep .. they are not overriding it so it is fine to do it ones in the base right ?

File: applications/MeshMovingApplication/custom_strategies/strategies/laplacian_meshmoving_strategy.h [hunk 5]
```diff
@@ -192,9 +192,6 @@ class LaplacianMeshMovingStrategy
       mstrategy_z->Solve();
     }
 
-    MoveMeshUtilities::MoveMesh(
```
Comment (philbucher) [inline applications/MeshMovingApplication/custom_strategies/strategies/laplacian_meshmoving_strategy.h:195]: @adityaghantasala this is what I mean, you should do the same in the Trilinos Stragegies

Comment (adityaghantasala) [reply applications/MeshMovingApplication/custom_strategies/strategies/laplacian_meshmoving_strategy.h:195]: Ok !

**Review by philbucher — APPROVED** [review]
now works :+1: 

@rubenzorrilla maybe you want to try as well?

**Comment by adityaghantasala** [comment]
@rubenzorrilla  did you test this already ?

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla did you test this already ?

I'll try ASAP.

**Comment by philbucher** [comment]
@dbaumgaertner @armingeiser I suggest you also take a look at this to make sure your things still work

**Comment by armingeiser** [comment]
> @dbaumgaertner @armingeiser I suggest you also take a look at this to make sure your things still work

Tests are running so it should be fine.

**Review by rubenzorrilla — APPROVED** [review]
I've tested the FluidDynamicsApplication as well as the FSIApplication. Both work.

**Comment by adityaghantasala** [comment]
Thank you for confirming guys .. I am merging this now !
